src/services/kratos/auth-device-account.ts

```diff
@@ -0,0 +1,249 @@
+        return new LikelyNoUserWithThisPhoneExistError(err.message || err)
+      }
+
+      if (err.message === "Request failed with status code 401") {
+        return new AuthenticationKratosError(err.message || err)
+      }
+
+      return new UnknownKratosError(err.message || err)
+    }
+  }
+
+  const createDeviceIdentity = async ({
+    deviceId,
+  }: {
+    deviceId: DeviceId
+  }): Promise<CreateKratosUserForDeviceSchemaResponse | KratosError> => {
+    const traits = { device: deviceId }
+    const identityBody: CreateIdentityBody = {
+      credentials: { password: { config: { password } } },
+      state: "active",
+      schema_id: "device_account_v0",
+      traits,
+    }
+    try {
+      const identityResults = await kratosAdmin.createIdentity({
+        createIdentityBody: identityBody,
+      })
+      console.log(identityResults.status)
+      const result = await loginDeviceAccount({ deviceId })
+      if (result instanceof KratosError) return result
+      const sessionToken = result.sessionToken
+      const kratosUserId = result.kratosUserId as UserId
+
+      const levelZeroAccountsConfig = getDefaultAccountsConfig()
+      levelZeroAccountsConfig.initialLevel = AccountLevel.Zero
+      const account = await createAccountForDeviceAccount({
+        userId: kratosUserId,
+        config: levelZeroAccountsConfig,
+        device: deviceId,
+      })
+      if (account instanceof Error) return account
+      return { sessionToken, kratosUserId }
+    } catch (err) {
+      if (err.message === "Request failed with status code 400") {
+        return new LikelyUserAlreadyExistError(err.message || err)
+      }
+
+      return new UnknownKratosError(err.message || err)
+    }
+  }
+
+  const upgradeToPhoneSchema = async ({
+    kratosUserId,
+    phone,
+  }: {
+    kratosUserId: UserId
+    phone: PhoneNumber
+  }) => {
+    let identity: Identity
+
+    try {
+      ;({ data: identity } = await kratosAdmin.getIdentity({ id: kratosUserId }))
+    } catch (err) {
+      if (err.message === "Request failed with status code 400") {
+        return new LikelyUserAlreadyExistError(err)
+      }
+
+      return new UnknownKratosError(err)
+    }
+
+    if (identity.schema_id !== "device_account_v0") {
+      return new IncompatibleSchemaUpgradeError()
+    }
+
+    // Save a history of the device to the kratos admin metadata
+    try {
+      // in kratos datetime format
+      const updated_at = getKratosDateNow()
+      if (identity.metadata_admin) {
+        // is array
+        if (Array.isArray(identity.metadata_admin)) {
+          identity.metadata_admin = [
+            ...identity.metadata_admin,
+            { device: identity.traits.device, updated_at },
+          ]
+        } else {
+          // is json object
+          identity.metadata_admin = [
+            identity.metadata_admin,
+            { device: identity.traits.device, updated_at },
+          ]
+        }
```

Comment:
do we need to maintain 2 scenario where it's an array OR an object?

currently we don't use `metadata_admin`, so if we device it should only be an object, we don't need this branching.

Reply:
i wanted to avoid accidentally over-writing any data in the metadata_admin, so i check for both types. i want the metadata_admin to be an array of object as a history log. just being extra cautious checking for existing data (even tho there should not be any)